alphagov/notifications-api
# Look at all previous days when sending letters

## What
https://www.pivotaltracker.com/n/projects/1443052

Previously, when running the `collate_letter_pdfs_for_day` task, we
would only send letters that were created between 5:30pm yesterday and
5:30 today.

Now we send letters that were created before 5:30pm today and that are
still waiting to be sent. This will help us automatically attempt to
send letters that may have fallen through the gaps and not been sent the
previous day when they should have been.

Previously we solved the problem of letters that had fallen the gap by
having to run the task with a date parameter for example
`collate_letter_pdfs_for_day('2020-02-18'). We no longer need this date
parameter as we will always look back across previous days too for
letters that still need sending.

Note, we have to change from using the pagination `list_objects_v2` to
instead getting each individual notification from s3. We reduce load by
using `HEAD` rather than `GET` but this will still greatly increase the
number of API calls. We acknowledge there will be a small cost to this,
say 50p for 5000 letters and think this is tolerable. Boto3 also handles
retries itself so if when making one of the many HEAD requests, there is
a networking blip then it should be retried automatically for us.

## How to test locally
Make sure you have sent a letter before 5:30pm that exists in the `development-letters-pdf` bucket for yesterdays print run.
Make sure you have sent a letter before 5:30pm that exists in  `development-letters-pdf` bucket for the print run 2 days ago.

Note, you can likely set this up by uploading two letters today. And then:
- change their created at date in your database
- move the file in the s3 bucket to the relevant days sending folder
- change the filename of the file you are moving in s3 to match your new created_at timestamp

```
source environment.sh
flask shell
from app.celery.letters_pdf_tasks import collate_letter_pdfs_for_day
collate_letter_pdfs_for_day()
```
